cpp/src/rolling/detail/range_utils.cuh

```diff
@@ -0,0 +1,869 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "rolling_utils.cuh"
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/rolling.hpp>
+#include <cudf/rolling.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/utilities/type_checks.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+#include <rmm/resource_ref.hpp>
+
+#include <cuda/functional>
+#include <cuda/std/limits>
+#include <cuda/std/type_traits>
+#include <thrust/binary_search.h>
+#include <thrust/copy.h>
+#include <thrust/distance.h>
+#include <thrust/execution_policy.h>
+
+#include <optional>
+#include <variant>
+
+namespace CUDF_EXPORT cudf {
+namespace detail {
+namespace rolling {
+
+/**
+ * @brief Indicates behaviour of boundary of a rolling window.
+ */
+enum class window_tag : std::int8_t {
+  BOUNDED_OPEN,    ///< Window is bounded by a value-based endpoint, endpoint is excluded.
+  BOUNDED_CLOSED,  ///< Window is bounded by a value-based endpoint, endpoint is included.
+  UNBOUNDED,       ///< Window runs to beginning (or end) of the group the row is in.
+  CURRENT_ROW,     ///< Window contains all rows that compare equal to the current row.
+};
+/**
+ * @brief A group descriptor for an ungrouped rolling window with nulls
+ *
+ * @param nulls_at_start Are the nulls at the start or end?
+ * @param num_rows The number of rows to be rolled over.
+ * @param null_count The number of nulls.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped
+ * iterator construction.
+ */
+struct ungrouped_with_nulls {
+  bool nulls_at_start;
+  cudf::size_type num_rows;
+  cudf::size_type null_count;
+
+  static constexpr bool has_nulls{true};
+  /**
+   * @copydoc ungrouped::row_info
+   */
+  [[nodiscard]] __device__ constexpr cuda::std::
+    tuple<size_type, size_type, size_type, size_type, size_type, size_type, size_type>
+    row_info(size_type i) const noexcept
+  {
+    if (nulls_at_start) {
+      return {null_count, 0, num_rows, 0, null_count, null_count, num_rows};
+    } else {
+      return {null_count,
+              num_rows,
+              null_count,
+              num_rows - null_count,
+              num_rows,
+              0,
+              num_rows - null_count};
+    }
+  }
```

Comment:
I should go back and try again, but I had what seemed to be miscompilation issues if I exposed all of these values via individual function calls instead of returning everything all at once.